content/chapters/software-stack/lab/README.md

```diff
@@ -0,0 +1,952 @@
+
+  Also, each lower layer provides a generic interface to the higher layer.
+  These generic interfaces "hides" possible differences in the even lower layers.
+  This way, a software stack ensures portability across different other parts of software (and hardware as well).
+  For example, the standard C library, that we will present shortly, ensures portability across different operating systems.
+
+TODO: Diagram of generic software stack
+
+TODO: Quiz
+
+## Modern Software Stacks
+
+Most modern computing systems use a software stack such as the one in the figure below:
+
+TODO: modern software stack
+
+This modern software stack allows fast development and provides a rich set of applications to the user.
+
+The basic software component is the **operating system** (OS) (technically the operating system **kernel**).
+The OS provides the fundamental primitives to interact with hardware (read and write data) and to manage the running of applications (such as memory allocation, thread creation, scheduling).
+These primitives form the **system call API** or **system API**.
+An item in the system call API, i.e. the equivalent of a function call that triggers the execution of a functionality in the operating system, is a **system call**.
+
+The system call API is well defined, stable and complete: it exposes the entire functionality of the operating system and hardware.
+However, it is also minimalistic with respect to features and it provides a low-level (close to hardware) specification, making it cumbersome to use and **not portable**.
+
+Due to the downsides of the system call API, a basic library, the **standard C library** (also called **libc**), is built on top of it.
+The standard C library wraps each system call into an equivalent function call, following a portable calling convention.
+Because the system call API uses an OS-specific calling convention, the standard C library typically wraps each system call into an equivalent function call, following a portable calling convention.
+More than these wraps, the standard C library provides its own API that is typically portable.
+Part of the API exposed by the standard C library is the **standard C API**, also called **ANSI C** or **ISO C**;
+this API is typically portable across all platforms (operating systems and hardware).
+Despite its name, the standard C library provides APIs that may not be standard, but particular to the underlying operating systems, such as system call wrapper functions.
+
+The existence of the standard C library is reliant on the C programming language, a very simple programming language and very close to the low-level view of the memory.
+Because of this, most higher-level and more feature rich programming languages rely on the C library.
+Each programming language typically provides a standard library of its own together with a runtime library, both of which rely on the C library.
```

Comment:
You can easily use D as an example of this. Not a demo, just a few sentences that D executables also link libc so the language doesn't have to reinvent the wheel. I'm not sure if this is needed though.